### DISCUSSION TITLE
GuildMember#joinedTimestamp being null

### DISCUSSION THREAD

**Discussion by TAEMBO** [discordjs/discord.js#10047] [Q&A]
All properties mentioned here may be assumed to be under an instance of the GuildMember class unless otherwise mentioned to help with readability.

In one of my TS projects, `joinedTimestamp` as well as `joinedAt` have a type that is either a number or Date respectively, or `null`. Why is that? I've looked on the Discord server, and every answer I saw regarding the reasoning for them being possibly null is because it's a partial GuildMember. With that knowledge, shouldn't the nullified values be strictly put in `PartialGuildMember` instead? The d.js typings for a PartialGuildMember have `joinedAt`, `joinedTimestamp`, and `pending` set to where they may be null already, so why is this nullish typing behavior extending into the full GuildMember structure?

Furthermore, in the full GuildMember structure typings, the `partial` getter has an explicit type definition of [false](https://github.com/discordjs/discord.js/blob/main/packages/discord.js/typings/index.d.ts#L1567), which I believe most would infer means that the GuildMember that they're working with is a full GuildMember structure - `joinedTimestamp` and `joinedAt` not being null. With that being said, the definition of this getter is `return this.joinedTimestamp === null;`, which solidifies that if `<GuildMember>.joinedTimestamp` is null, then `<GuildMember>.partial` is true, and if it isn't null, then `partial` should be false. If the type definition for the `partial` getter in a full GuildMember structure is explicitly `false`, why is `joinedTimestamp` and `joinedAt` still being typed as if it can be null when the `partial` type definition says otherwise?

**Comment by Jiralite** [comment]
Receiving a `PRESENCE_UPDATE` event on an uncached guild member has no `joined_at` field (or guild member data as a whole), resulting in this being `null`.

**Reply by TAEMBO** [reply]
Ok, gotcha. With knowing that, that would mean that GuildMember#partial would be `true` in that instance as opposed to its type definition of `false`, yes? That seems to still leave typings that don't properly reflect their runtime values unless there's something else I'm missing. If this debacle is solely due to the presence update event, would it be better to label the new data as a union type of a GuildMember or PartialGuildMember instead of just GuildMember, and then remove the nullish option from these aforementioned properties from the full GuildMember typings?
